core/recipe/runner.go

```diff
@@ -30,192 +30,150 @@ func NewRunner(
 		monitor = new(defaultMonitor)
 	}
 	return &Runner{
-		extractorFactory:        extractorFactory,
+		extractorFactory: extractorFactory,
 		processorFactory: processorFactory,
 		sinkFactory:      sinkFactory,
 		monitor:          monitor,
 	}
 }
 
-func (r *Runner) Run(recipe Recipe) (err error) {
-	r.wg = new(sync.WaitGroup)
-	r.errChan = make(chan error)
-	getDuration := r.startDuration()
-	success := true
+func (r *Runner) Run(recipe Recipe) (run Run) {
+	var wg sync.WaitGroup
+	var (
+		getDuration = r.startDuration()
+		channel     = make(chan interface{})
+	)
+	run.Recipe = recipe
 
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
+	// run extractors
+	{
+		outChannel := channel
+		go func() {
+			err := r.runExtractor(ctx, recipe.Source, outChannel)
+			if err != nil {
+				run.Error = r.buildTaskError(TaskTypeExtract, recipe.Source.Type, err)
+			}
 
-	// run all sinks
-	var sinkOuts []chan interface{}
-	for _, sink := range recipe.Sinks {
-		task := &Task{
-			Type:   TaskTypeSink,
-			Status: TaskStatusReady,
-			Name:   sink.Name,
-			Config: sink.Config,
-		}
-		r.wg.Add(1)
-		currentSink := make(chan interface{})
-		sinkOuts = append(sinkOuts, currentSink)
-
-		go r.runSink(ctx, task, currentSink)
-	}
-
-	// run extract
-	extractOut := make(chan interface{})
-	extractTask := &Task{
-		Type:   TaskTypeExtract,
-		Status: TaskStatusReady,
-		Name:   recipe.Source.Type,
-		Config: recipe.Source.Config,
+			close(outChannel)
+		}()
 	}
-	r.wg.Add(1)
-	go r.runExtractor(ctx, extractTask, extractOut)
 
 	// run processors
-	processIn := extractOut
-	var processorOuts []chan interface{}
-	for _, processor := range recipe.Processors {
-		r.wg.Add(1)
-		task := &Task{
-			Type:   TaskTypeProcess,
-			Status: TaskStatusReady,
-			Name:   processor.Name,
-			Config: processor.Config,
-		}
-		processOut := make(chan interface{})
-		processorOuts = append(processorOuts, processOut)
-		go r.runProcessor(ctx, task, processIn, processOut)
-		processIn = processOut
-	}
+	for _, processorRecipe := range recipe.Processors {
+		inChannel := channel
+		outChannel := make(chan interface{})
+
+		// need to store the recipe since it would be needed inside a goroutine
+		// not storing it inside the loop scope would cause
+		// processorRecipe to always be the last recipe in the loop
+		tempRecipe := processorRecipe
+		go func() {
+			err := r.runProcessor(ctx, tempRecipe, inChannel, outChannel)
+			if err != nil {
+				run.Error = r.buildTaskError(TaskTypeProcess, tempRecipe.Name, err)
+			}
 
-	go func() {
-		// fan out service for sinks
-		for val := range processIn {
-			for _, so := range sinkOuts {
-				so <- val
+			close(outChannel)
+		}()
+
+		// replace the channel with the new out channel
+		// this would allow the next processor or sink to
+		// receive the processed data instead of data directly from extractor
+		channel = outChannel
+	}
+
+	// run sinks
+	for _, sinkRecipe := range recipe.Sinks {
```

Comment:
Should we run sinks first. Because if sinks are not ready there is no point of running the extractors. Order should be sinks > processors > extractors.

Reply:
we can't due to channels dependencies in this current logic.

say we have a recipe with 1 extractor (`extr-a`), 2 processors (`proc-a` & `proc-b`), and 2 sinks (`sink-a`, `sink-b`).
then we would have below structure:

```
extr-a --> channel-1 --> proc-a
proc-a --> channel-2 --> proc-b
proc-b --> channel-3 --> [sink-a, sink-b]
```

so basically I need to define the inputter of the channel before the listener of the channel.

technically the order of the execution won't matter much since they are all running asynchronously, so it is totally possible to setup sinks first.
but it would make the channels management to be much much harder.

> Because if sinks are not ready there is no point of running the extractors

yes, but maybe we can also look it this way, `if there is no extractor, there is nothing to be done and to listen to`.
Even processors would be useless without extractors. Also `extractor -> processors -> sinks` is actually what is going on. wdyt?

Reply:
I think it is fine for now. Later we can think of having an init function in each extractor and processor that can run first before starting extraction to check all connections etc.